### PR TITLE
Patch lint attribute spacing

### PR DESCRIPTION
**Pull request by siva-sundar**
Implements https://github.com/rwjblue/ember-template-lint/issues/28

### CODE REVIEW

**Comment by siva-sundar** [comment]
Implements https://github.com/rwjblue/ember-template-lint/issues/28

File: lib/rules/lint-attribute-indentation.js [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+
+const getWhiteSpaceLength = function(statement) {
+  let whiteSpace = statement.match(/^\s+/) || [];
+  return (whiteSpace[0] || '').length;
+};
+
+module.exports = class AttributeSpacing extends Rule {
+
+  getBlockParamStartLoc(node) {
+    let actual, expected;
```
Comment (siva-sundar) [inline lib/rules/lint-attribute-indentation.js]: Is there a direct way to get the location of block params?

File: lib/rules/lint-attribute-indentation.js [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+        {{contact-details
+          age
+          firstName=firstName
+          fullName=fullName
+        }}
+    */
+    let expectedColumnStart = node.loc.start.column + 2; //params should be after 2 positions from component start node
+    let expectedLineStart = node.loc.start.line + 1;
+
+    this.iterateParams(node.params, 'positional', expectedLineStart, expectedColumnStart, node);
```
Comment (siva-sundar) [inline lib/rules/lint-attribute-indentation.js]: Current implementation enforces the `positional params` to follow the same indentation rule.

Comment (rwjblue) [reply lib/rules/lint-attribute-indentation.js]: Seems good to me (I think)....

File: test/unit/rules/lint-attribute-indentation-test.js [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    //Block form with no params
+    '{{#contact-details' + '\n' +
+    'as |contact|}}' + '\n' +
+    '  {{contact.fullName}}' + '\n' +
+    '{{/contact-details}}'
+  ],
+
+  bad: [{
+      //MustacheStatement
+    template: '{{contact-details firstName=firstName lastName=lastName}}',
```
Comment (rwjblue) [inline test/unit/rules/lint-attribute-indentation-test.js]: I think the rule needs to be a bit nuanced here.  I think this should be a "good" example (non-block single line invocation with less than 80 chars).

File: test/unit/rules/lint-attribute-indentation-test.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+      'column': 55,
+      'line': 1,
+      'message': 'Incorrect indentation of close curly braces \'}}\' for the component \'{{contact-details}}\' beginning at L1:C55. Expected to be indentation at L2:C0 with an of 0 but was found at 55',
+      'moduleId': 'layout.hbs',
+      'rule': 'attribute-indentation',
+      'source': '{{contact-details firstName=firstName lastName=lastName}}'
+    }]
+  }, {
+      //MustacheStatement with a positional param
+    template: '{{contact-details firstName lastName=lastName}}',
```
Comment (rwjblue) [inline test/unit/rules/lint-attribute-indentation-test.js]: Same as above, I think this should be a good example.

**Comment by siva-sundar** [comment]
Thank you. I have added an option to configure the length of the open invocation statement. Also made the changes you have mentioned.

File: test/unit/rules/lint-attribute-indentation-test.js [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    '{{contact-details' + '\n' +
+    '  firstName=firstName' + '\n' +
+    '  lastName=lastName' + '\n' +
+    '}}',
+    //Non Block form with no params
+    '{{contact-details}}',
+    //Non Block form within 70 characters
+    {
+      config: {
+        'open-invocation-max-len': 70
```
Comment (rwjblue) [inline test/unit/rules/lint-attribute-indentation-test.js]: Just to be clear, this one would pass even without custom config right?

Comment (siva-sundar) [reply test/unit/rules/lint-attribute-indentation-test.js]: Yes, I'll change the example to have more than the default length(80 chars).

**Comment by siva-sundar** [comment]
I have added few more test cases. Can you take a look at it?

**Comment by rwjblue** [comment]
Awesome, thank you very much for all of your hard work on this!

File: test/unit/rules/lint-attribute-indentation-test.js [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+      config: {
+        'open-invocation-max-len': 120
+      },
+      template: '{{contact-details firstName=firstName lastName=lastName avatarUrl=avatarUrl age=age address=address phoneNo=phoneNo}}'
+    },
+    //Open-invocation with multiple lines.
+    '{{contact-details' + '\n' +
+    '  firstName=firstName' + '\n' +
+    '  lastName=lastName' + '\n' +
+    '}}',
```
Comment (amk221) [inline test/unit/rules/lint-attribute-indentation-test.js]: Will attribute-indentation support not having to have }} on its own line. I like to indent over multiple lines, but retain the }} at the end

Comment (siva-sundar) [reply test/unit/rules/lint-attribute-indentation-test.js]: We don't have an option to specify this style of indentation. Perhaps, we could create a sub rule?

cc: @rwjblue

Comment (amk221) [reply test/unit/rules/lint-attribute-indentation-test.js]: Worth a though, as I think this should be supported to:
```handlebars
{{my-component
  foo='bar'
  baz='qux'
  my-attr=(component 'my-other-component'
    foo='bar'
    baz='qux')}}
```

Comment (rwjblue) [reply test/unit/rules/lint-attribute-indentation-test.js]: for single line invocation, closing at the end of the last line seems ok (and we can update this rule to allow it).

@amk221 - Can you create an issue to track this?

**Pull request merged**